data/src/history.rs

```diff
@@ -397,6 +449,90 @@ impl History {
     }
 }
 
+/// Insert the incoming message into the provided vector, sorted
+/// on server time
+///
+/// Deduplication is only checked +/- 1 second around the server time
+/// of the incoming message. Either message IDs match, or server times
+/// have an exact match + target & content.
+pub fn insert_message(messages: &mut Vec<Message>, message: Message) {
```

Comment:
Didn't we have a commit which used a binary search implementation for inserting messages?

Reply:
Yes, this is the `insert_message` from that work we did together, completely unchanged.  I was lazy about persisting the commit history while refactoring, so it's all been squashed into one commit.

The one thing this doesn't keep related to that work is the sorting of saved histories on load.  In that version of the `chathistory` PR, I had changed the history storage scheme to no longer be obfuscated (no longer use hashes in naming the files).  If histories were not found with the non-hashed storage scheme (where histories can be expected to be sorted already), then it would attempt to load & sort a history from the old, hashed storage scheme.  I still like that idea, but it seemed like this PR might have too many changes already, so I omitted it this time around.